runtime/src/iree/hal/drivers/hip/hip_driver.c

```diff
@@ -465,10 +472,120 @@ static iree_status_t iree_hal_hip_driver_create_device_by_index(
+    return status;
+  }
+  status = iree_hal_hip_driver_get_device_id_by_index(base_driver, device_path,
+                                                      &device_id);
+  if (iree_status_is_ok(status)) {
+    *out_id = device_id;
+    return iree_ok_status();
+  }
+  if (!iree_status_is_invalid_argument(status)) {
+    return status;
+  }
+  return iree_make_status(IREE_STATUS_UNIMPLEMENTED, "unsupported device path");
+}
+
+static iree_status_t iree_hal_hip_driver_create_multi_queue_device_by_ids(
+    iree_hal_driver_t* base_driver, iree_hal_device_id_t* device_ids,
+    iree_host_size_t device_count, iree_host_size_t param_count,
+    const iree_string_pair_t* params, iree_allocator_t host_allocator,
+    iree_hal_device_t** out_device) {
+  IREE_ASSERT_ARGUMENT(base_driver);
+  IREE_ASSERT_ARGUMENT(out_device);
+
+  iree_hal_hip_driver_t* driver = iree_hal_hip_driver_cast(base_driver);
+  IREE_TRACE_ZONE_BEGIN(z0);
+
+  // Ensure HIP is initialized before querying it.
+  IREE_RETURN_AND_END_ZONE_IF_ERROR(z0, iree_hal_hip_init(driver));
+
+  hipDevice_t* devices;
+  IREE_RETURN_IF_ERROR(iree_allocator_malloc(
+      host_allocator, sizeof(*devices) * device_count, (void**)&devices));
+
+  for (iree_host_size_t i = 0; i < device_count; ++i) {
+    if (device_ids[i] == IREE_HAL_DEVICE_ID_DEFAULT) {
+      return iree_make_status(
+          IREE_STATUS_INVALID_ARGUMENT,
+          "Invalid to create a device group with the default device id");
+    } else {
+      devices[i] = IREE_DEVICE_ID_TO_HIPDEVICE(device_ids[i]);
+    }
+  }
+
+  iree_string_view_t device_name = iree_make_cstring_view("hip");
+
+  // Attempt to create the device now.
+  iree_status_t status = iree_hal_hip_device_create(
+      base_driver, device_name, &driver->device_params, &driver->hip_symbols,
+      &driver->nccl_symbols, device_count, devices, host_allocator, out_device);
+  iree_allocator_free(host_allocator, devices);
+
+  IREE_TRACE_ZONE_END(z0);
+  return status;
+}
+
+static iree_status_t iree_hal_hip_driver_create_multi_queue_device_by_path(
+    iree_hal_driver_t* base_driver, iree_string_view_t driver_name,
+    iree_string_view_t device_path, iree_host_size_t param_count,
+    const iree_string_pair_t* params, iree_allocator_t host_allocator,
+    iree_hal_device_t** out_device) {
+  IREE_ASSERT_ARGUMENT(base_driver);
+  IREE_ASSERT_ARGUMENT(out_device);
+
+  uint64_t multi_count = 0;
+  for (iree_host_size_t offs = 0; offs < device_path.size;) {
+    iree_host_size_t comma_pos =
+        iree_string_view_find_char(device_path, ',', offs);
+    if (comma_pos == IREE_STRING_VIEW_NPOS) {
+      comma_pos = device_path.size;
+    }
+    offs = comma_pos + 1;
+    multi_count++;
+  }
+
+  iree_hal_device_id_t* device_ids;
+  IREE_RETURN_IF_ERROR(iree_allocator_malloc(
+      host_allocator, sizeof(*device_ids) * multi_count, (void**)&device_ids));
+
+  iree_host_size_t idx = 0;
+  for (iree_host_size_t offs = 0; offs < device_path.size;) {
```

Comment:
```suggestion
  for (iree_host_size_t offset = 0; offs < device_path.size;) {
```
(if using non-i/j/k then use real words)